src/modules/performance.ts

```diff
@@ -0,0 +1,360 @@
+  public readonly spans: Span[] = [];
+
+  /**
+   * Constructor for Transaction
+   *
+   * @param data - Data to initialize the transaction with. Contains id, name, startTime, tags
+   * @param performance - Reference to the PerformanceMonitoring instance that created this transaction
+   */
+  constructor(
+    data: Omit<Transaction, 'startSpan' | 'finish'>,
+    private readonly performance: PerformanceMonitoring
+  ) {
+    Object.assign(this, data);
+  }
+
+  /**
+   *
+   * @param name
+   * @param metadata
+   */
+  public startSpan(name: string, metadata?: Record<string, unknown>): Span {
+    const data = {
+      id: id(),
+      transactionId: this.id,
+      name,
+      startTime: getTimestamp(),
+      metadata,
+    };
+
+    const span = new Span(data);
+
+    this.spans.push(span);
+
+    return span;
+  }
+
+  /**
+   *
+   */
+  public finish(): void {
+    // Finish all unfinished spans
+    this.spans.forEach(span => {
+      if (!span.endTime) {
+        span.finish();
+      }
+    });
+
+    this.endTime = getTimestamp();
+    this.duration = this.endTime - this.startTime;
+    this.performance.queueTransaction(this);
+  }
+}
+
+/**
+ * Class representing a sampled out transaction that won't be sent to server
+ */
+class SampledOutTransaction extends Transaction {
+  /**
+   * Constructor for SampledOutTransaction
+   *
+   * @param data - Data to initialize the transaction with. Contains id, name, startTime, tags and spans
+   */
+  constructor(data: Omit<Transaction, 'startSpan' | 'finish'>) {
+    super(data, null as unknown as PerformanceMonitoring); // performance не используется
+  }
+
+  /**
+   * Start a new span within this sampled out transaction
+   *
+   * @param name - Name of the span
+   * @param metadata - Optional metadata to attach to the span
+   * @returns A new Span instance that won't be sent to server
+   */
+  public startSpan(name: string, metadata?: Record<string, unknown>): Span {
+    const data = {
+      id: id(),
+      transactionId: this.id,
+      name,
+      startTime: getTimestamp(),
+      metadata,
+    };
+
+    const span = new Span(data);
+
+    this.spans.push(span);
+
+    return span;
+  }
+
+  /**
+   *
+   */
+  public finish(): void {
+    // Do nothing - don't send to server
+  }
+}
+
+/**
+ * Class for managing performance monitoring
+ */
+export default class PerformanceMonitoring {
+  /**
+   * Timer for batch sending
+   */
+  private batchTimeout: ReturnType<typeof setTimeout> | null = null;
+
+  /**
+   * Map of active transactions by their ID
+   * Used to:
+   * - Track transactions that haven't been finished yet
+   * - Finish all active transactions on page unload/process exit
+   * - Prevent memory leaks by removing finished transactions
+   */
+  private activeTransactions: Map<string, Transaction> = new Map();
+
+  /**
+   * Queue for transactions waiting to be sent
+   */
+  private sendQueue: Transaction[] = [];
```

Comment:
Questionable:

maybe we should use Stack instead of array to limit the size of data being sent?
```suggestion
  private sendQueue: Stack<Transaction> = new Stack({size: 40});
```